File: lintcode/golang/1165_subtree_of_another_tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

/**
 * @param s: the s' root
 * @param t: the t' root
 * @return: whether tree t has exactly the same structure and node values with a subtree of s
 */
func isSubtree (s *TreeNode, t *TreeNode) bool {
    // Write your code here
    if s == nil {
        return t == nil
    }
    
    if s.Val == t.Val && isSame(s, t) {
        return true
    }
    
    return isSubtree(s.Left, t) || isSubtree(s.Right, t)
}

func isSame (s *TreeNode, t *TreeNode) bool {
    if s == nil {
        return t == nil
    }
    
    if t == nil || s.Val != t.Val {
        return false
    }
    
    return isSame(s.Left, t.Left) && isSame(s.Right, t.Right)
}

File: lintcode/python/1165_subtree_of_another_tree.py
#!/usr/bin/python -t

# bst and dfs

"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""

class Solution:
    """
    @param s: the s' root
    @param t: the t' root
    @return: whether tree t has exactly the same structure and node values with a subtree of s
    """
    def isSubtree(self, s, t):
        # Write your code here
        if not t:
            return not s
        if not s:
            return not t

        if s.val == t.val and self.isSame(s, t):
            return True

        return self.isSubtree(s.left, t) or self.isSubtree(s.right, t)

    def isSame(self, s, t):
        if not s:
            return not t

        if t == None or s.val != t.val:
            return False

        return self.isSame(s.left, t.left) and self.isSubtree(s.right, t.right)
